### PR TITLE
Flip insecure defined flask session config

### PR DESCRIPTION
**Pull request by clavedeluna**
## Overview
*A codemod that will check if `SESSION_COOKIE_HTTPONLY`, `SESSION_COOKIE_SECURE` or `SESSION_COOKIE_SAMESITE` are set to an insecure value on a flask app config object.*

## Description

* This codemod checks any instances of a Flask app for any calls to its configuration. If any of those calls leads to setting any of the 3 flags to their unsafe value, this codemod will change the value.
* Note that after quite a bit of back and forth, we've decided this codemod will not at this time add set the secure values for any of these configs. We will only handle the case when the user specifically set to an unsafe value. We're aware that some default values are unsafe. 
* I've left some commented out code which did handle ^ above by adding a line at the very end of the module any of the config values which were not set. We decided there is some risk to adding this line at the end of the module so we won't do so right now.

### CODE REVIEW

File: src/core_codemods/secure_flask_session_config.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        self.flask_app_name = ""
+        # Later: if we want to store configs to write later
+        # self.configs_to_write = self.SECURE_SESSION_CONFIGS.copy()
+        self.file_context = file_context
+
+    def _store_flask_app(self, original_node) -> None:
+        flask_app_parent = self.get_metadata(ParentNodeProvider, original_node)
+        match flask_app_parent:
+            case cst.AnnAssign() | cst.Assign():
+                flask_app_attr = flask_app_parent.targets[0].target
```
Comment (andrecsilva) [inline src/core_codemods/secure_flask_session_config.py]: This won't work with `AnnAssign`, as the target value is stored in the `target` attribute. You should separate the cases.

This may help:
```python
    def _extract_targets_of_assignment(
        self, assignment: cst.AnnAssign | cst.Assign | cst.WithItem | cst.NamedExpr
    ) -> list[cst.BaseExpression]:
        match assignment:
            case cst.AnnAssign():
                if assignment.target:
                    return [assignment.target]
            case cst.Assign():
                return [t.target for t in assignment.targets]
            case cst.NamedExpr():
                return [assignment.target]
            case cst.WithItem():
                if assignment.asname:
                    return [assignment.asname.name]
        return []
```

File: src/core_codemods/secure_flask_session_config.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                self.flask_app_name = flask_app_attr.value
```
Comment (andrecsilva) [inline src/core_codemods/secure_flask_session_config.py]: This assumes that the target is a `Name`, but the target may be any slew of different types of expressions like `Subscript`:(`l[0] = Flask(...)`) , `Attribute`:(`a.b = Flask(...)`), etc. So you should make it fail in the other cases.

My suggestion is to break this detection part into a new method `is_assigned_to_variable` with return `Optional[Name]`.

File: src/core_codemods/secure_flask_session_config.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        app_config_node = cst.Attribute(value=app_name, attr=config)
+        update = cst.Name(value="update")
+        return matchers.matches(
+            original_node.func, matchers.Attribute(value=app_config_node, attr=update)
+        )
+
+    def _is_config_subscript(self, original_node: cst.Assign):
+        config = cst.Name(value="config")
+        app_name = cst.Name(value=self.flask_app_name)
+        app_config_node = cst.Attribute(value=app_name, attr=config)
```
Comment (andrecsilva) [inline src/core_codemods/secure_flask_session_config.py]: When using `matchers.matches`, you should use the `matchers` version of the nodes instead (i.e. `matchers.Name` instead of `cst.Name`).

See: https://libcst.readthedocs.io/en/latest/matchers.html#libcst.matchers.matches

File: src/core_codemods/secure_flask_session_config.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                safe_value = self._get_secure_config_val(key)
+                self.report_change(original_node)
+                return updated_node.with_changes(value=safe_value)
+        return updated_node
+
+    def _is_config_update_call(self, original_node: cst.Call):
+        config = cst.Name(value="config")
+        app_name = cst.Name(value=self.flask_app_name)
+        app_config_node = cst.Attribute(value=app_name, attr=config)
+        update = cst.Name(value="update")
```
Comment (andrecsilva) [inline src/core_codemods/secure_flask_session_config.py]: See comment in `_is_config_subscript`. Same issue.

**Review by andrecsilva — CHANGES_REQUESTED** [review]
Some minor issues, otherwise it looks good.

File: src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py
Comment (drdavella) [inline src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py:1]: I think this file doesn't belong in this PR, it looks like it's from another ticket.

**Review by drdavella — APPROVED** [review]
Thanks for all of the back and forth on this, I know it ended up being fairly tricky. 

Based on the expected behavior in the tests this looks good to me. My only substantive comment is to remove the file that doesn't belong in this PR. Once @andrecsilva's other comments are addressed this should be good to go.

File: src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py
Comment (clavedeluna) [reply src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py:1]: This change does belong here, I didn't add the file, I just moved a function to utils so I had to change the import

File: src/core_codemods/secure_flask_session_config.py (hunk 1 above)
Comment (clavedeluna) [reply src/core_codemods/secure_flask_session_config.py]: I assume you wrote that function and it isn't from somewhere I should cite, so I'm taking it as is.

File: src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py
Comment (drdavella) [reply src/codemodder/project_analysis/file_parsers/setup_py_file_parser.py:1]: Ah sorry my bad.

File: src/core_codemods/secure_flask_session_config.py (hunk 1 above)
Comment (andrecsilva) [reply src/core_codemods/secure_flask_session_config.py]: I wrote it for the file resource leak codemod. This PR will probably be merged before, so go ahead.

**Review by andrecsilva — APPROVED** [review]
